### DISCUSSION TITLE
client cannot find hot reload files (possibly a proxy problem)

### DISCUSSION THREAD

**Discussion by raquelhortab** [webpack/webpack#16425] [Q&A]
Hi, I got a CRA dev app running in a docker container. Locally, webpack's hot reload works fine, but in production, it does not.

This is the scenario for my local configuration:

- local docker container with the app
- I start the app it using `WDS_SOCKET_PORT=$WDS_SOCKET_PORT npm start` where WDS_SOCKET_PORT is the port assigned by docker
- Both error overlay and hot-reload work fine

In production, webpack hot reload does not work, this is my scenario:
- We use a proxy that maps urls like `<ourserver>/services-node-xxxxx`, to the corresponding docker port (being xxxxx  the port). 
- The production docker container runs the app using `WDS_SOCKET_PORT=0 WDS_SOCKET_PATH="${PUBLIC_URL}/ws" PUBLIC_URL=$PUBLIC_URL npm start`. Here, `WDS_SOCKET_PORT=0` is so the error overlay works, `PUBLIC_URL` corresponds to `/services-node-xxxxx`
- The production container has a setupProxy.js: 
```
module.exports = function(app) {
        app.use(
            '/static',
            createProxyMiddleware({
                target: 'http://localhost:3000',
                pathRewrite:{
                    '/static':`${process.env.PUBLIC_URL}/static`
                }
            })
        );
};
```
Otherwise, the client could not find static files.

I have been trying to debug what is making hot reload not work and I've narrowed it down to the fact that the requests like `main.131aa0252160dc31a717.hot-update.json` are not found. The client does the request, and I can see in the server console logs that the file has been generated (so the file watcher works), but the client cannot retrieve it. I guess it has to do with some of the two proxies or maybe some env variable although the requests I see seem logical.

Here are some videos to make it clearer. 
First, the working version running locally:
[screencast-nimbus-capture-2022.10.27-16_50_57.webm](https://user-images.githubusercontent.com/19463905/198328160-78f9854d-2834-4242-a8af-489294c6b422.webm)

Here, the production version that fails (note that the request status is 200OK but it is actually getting an html because it did not find the json file). Also note that WS connection does work, since the error overlay appears automatically whenever I generate an error.
[screencast-nimbus-capture-2022.10.27-16_39_02 (1).webm](https://user-images.githubusercontent.com/19463905/198328512-c655f855-a14e-4d9e-81fa-cfc1d2535931.webm)

**Comment by raquelhortab** [comment]
I've been trying to deal with this for quite a while so if anyone could be so nice to help I would not mind posting this on StackOverflow to give credit or even "buymeacoffee-them"

**Reply by alexander-akait** [reply]
I think you need rewrite request for js/css/etc files too, you can debug URL here https://github.com/webpack/webpack-dev-middleware/blob/master/src/utils/getFilenameFromUrl.js#L53, if you don't reach this place it means something blocked your request before

**Reply by raquelhortab** [reply]
Okay so I am seeing something here! Here's my debugging:
```
getFilenameFromUrl called
/main.c3a315512688edee7712.hot-update.json (url)
paths
[
  {
    outputPath: '/home/react_app/build',
    publicPath: '/services-node-44248'
  }
]
/main.c3a315512688edee7712.hot-update.json
urlObject
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: '/main.c3a315512688edee7712.hot-update.json',
  path: '/main.c3a315512688edee7712.hot-update.json',
  href: '/main.c3a315512688edee7712.hot-update.json'
}
/main.c3a315512688edee7712.hot-update.json
publicPathObject
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: '/services-node-44248',
  path: '/services-node-44248',
  href: '/services-node-44248'
}
```

It does not go into 
```
if (urlObject.pathname && urlObject.pathname.startsWith(publicPathObject.pathname))
```
because `urlObject.pathname` (/main.c3a315512688edee7712.hot-update.json) does not start with `publicPathObject.pathname` (/services-node-44248).

In the local version it does find it because `publicPathObject.pathname` is `'/'`.

Now, if the url is `/main.c3a315512688edee7712.hot-update.json` where is it getting the `/services-node-44248` path from? is it using the env variable `PUBLIC_URL`? I could proxy any request from `/PUBLIC_URL` to `/` but I really cannot, because it seems to conflict with the `/static` proxy rule. @alexander-akait any idea?

thanks for the help so far

**Reply by raquelhortab** [reply]
I'm currently looking around webpackDevServer.config.js and webpack.config.js in react-scripts

**Reply by raquelhortab** [reply]
I see, in paths.js there's 
```
publicUrlOrPath = getPublicUrlOrPath(
  process.env.NODE_ENV === 'development',
  require(resolveApp('package.json')).homepage,
  process.env.PUBLIC_URL
);
```

So it does use PUBLIC_URL

**Reply by raquelhortab** [reply]
While I don't find a better approach I might just use patch-package to make this not use the PUBLIC_URL env variable (update: not really a good solution since it can mess up with other requests)

**Reply by alexander-akait** [reply]
Yeah, it is not good, but I think CRA was not design for that usage... You can try to proxy files which ends on `js`/`css`/etc (i.e. static) + websockets from dev server, it should work, maybe you need think about URL structure and refactor it to avoid overlaps

**Reply by raquelhortab** [reply]
I still can't get it to work...  The root of the problem is:
- I need to proxy any request like `myserver/services-node-xxxxx/whatever` to `mydockerserver:xxxxx/whatever`
- Without any configuration, CRA  and webpack generate an html file where all paths are relative to `/`, so any static files are searched in `myserver/static` instead of `myserver/services-node-xxxxx/static`
- Now, there's the PUBLIC_URL option that makes the paths in the html apparently correct (using `PUBLIC_URL=/services-node-xxxxx/` urls in html are now `/services-node-xxxxx/static`). However, it doesn't only change the paths in the public files but also where they are served in the docker/app itself. So now, `myserver/services-node-xxxxx/whatever` is proxied to `mydockerserver:xxxxx/whatever` by nginx, so the requests reaches the app as `localhost:3000/whatever` but because of PUBLIC_PATH,  the app is actually serving things to `localhost:3000/services-node-xxxxx/whatever` so things are not correctly found.

The real problem then is, isn't there any webpack option similar to PUBLIC_URL that allows setting the actual public urls in public files but does not change how the app serves the files?

**Reply by raquelhortab** [reply]
it seems it could be enough if relative public_url were allowed since setting `publicPath: ''` in `node_modules/react-scripts/config/webpack.config.js` seems to fix the issue. The problem is I'd need an official workaround (through environment variables) rather than a patch...
